numpy/lib/shape_base.py

```diff
@@ -90,7 +90,13 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
     i.put(indlist, ind)
     res = func1d(arr[tuple(i.tolist())], *args, **kwargs)
     #  if res is a number, then we have a smaller output array
-    if isscalar(res):
+    asscalar = isscalar(res)
```

Comment:
I think it would be simpler to just do
```
if isscalar or not hasattr(res, `__len__`):
```
Even so, for better behavior I'd probably save the results of applying `func` in a list, make a array out of that, and then reshape it. One could instead maybe  do `asarray(res)` up front .and check the dimension of that. I think that would solve the problem here and also work for functions that, say, return a set constructed from the elements of the axis and that currently fail.